kvbc/src/v4blockchain/detail/latest_keys.cpp

```diff
@@ -490,4 +504,132 @@ bool LatestKeys::LKCompactionFilter::Filter(int /*level*/,
   return true;
 }
 
+// when we get a range delete, we can't perform the operation on the same block since we can't recover from it
+//(we can't know which keys were deleted and recover them).
+// therefore, we add a range delete marker to storage.
+// This method applies the range delete markers and should be called on a stable state
+// e.g. the next BFT sequence number.
+// Furthermore, as applying the range delete might occur a few blocks after its addition, it can happen that it will
+// delete updates with a higher version, therefore we iterate over the blocks that were added after,
+// and restore any update that is within the deletion range.
+void LatestKeys::ApplyRangeDelete(v4blockchain::detail::Blockchain* block_chain) {
+  auto itr = native_client_->getIterator(v4blockchain::detail::LATEST_KEYS_CF);
+  itr.seekAtLeast(RANGE_DELETE_PREFIX);
+  if (!itr || itr.keyView().substr(0, 1) != RANGE_DELETE_PREFIX) {
+    return;
+  }
+  auto write_batch = native_client_->getBatch();
+  // iterate over the range delete keys
+  while (itr && itr.keyView().substr(0, 1) == RANGE_DELETE_PREFIX) {
+    // key prefix is |range delete prefix|version|category prefix|...
+    ConcordAssertGT(itr.keyView().size(), VERSION_SIZE + RANGE_DELETE_PREFIX.size() + 1);
+    ConcordAssertGT(itr.valueView().size(), 1);
+    auto range_pair = extractRange(itr);
+    auto start_range = range_pair.first;
+    auto end_range = range_pair.second;
+    // category information
+    auto category_prefix = std::string(itr.keyView().substr(1 + VERSION_SIZE, 1).data(), 1);
+    auto category_id = getCategoryFromPrefix(category_prefix);
+
+    auto version = concordUtils::fromBigEndianBuffer<BlockId>(itr.keyView().substr(1, VERSION_SIZE).data());
+    write_batch.delRange(v4blockchain::detail::LATEST_KEYS_CF, start_range, end_range);
+    write_batch.del(v4blockchain::detail::LATEST_KEYS_CF, itr.keyView());
+    LOG_DEBUG(V4_BLOCK_LOG,
+              "Applying delete-range, start " << start_range << " ,end " << end_range << " version " << version);
+    itr.next();
+    // here we should iterate over the blocks with higher version than the range-delete
+    // and look for keys that are within the range to re-add them.
```

Comment:
As RocksDB range delete only accepts start-end , it doesn't iterate over the KVs explicitly and delete them but adds a marker in its sst files. 
So it will cover any key that is within the range, I couldn't think of any way to prevent it other than re-writing it